### PR TITLE
18779 Use Vuetify 3 DatePicker

### PR DESCRIPTION
**Pull request by leodube-aot**
*Issue #:* /bcgov/entity#18779

*Description of changes:*
- Updated Vuetify 3 to version with DatePicker component
- Updated our DatePicker component to use the new one from Vuetify 3. The storybook stories seem to all work the same as our previous DatePicker
- Updated storybook files
- I'll rebase and ~finish up unit tests~ _(tech debt ticket created instead)_ once Ketaki's PR (#212 ) is merged as she has fixed up the unit test setup

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [inline src/components/date-picker/DatePicker.vue:20]: The `v-menu` would not open with `v-on` and `on`. Using `props` seems to work

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [inline src/components/date-picker/DatePicker.vue:35]: `name` expects a string

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [inline src/components/date-picker/DatePicker.vue:50]: The Vuetify 3 DatePicker does not seem to be working with `string` dates, so I switched the private `dateText` variable to be `data` with type `Date`.

File: src/mixins/date-mixin.ts
Comment (leodube-aot) [inline src/mixins/date-mixin.ts:72]: Since Vuetify 3 Datepicker returns a `Date`, it also includes timezone information and defaults the time to `00:00:00`. Our previous datepicker would ignore this since it would return a string. 

I tested the Vuetify 3 DatePicker for non-pacific timezones and this resulted in bizarre behaviour. Notice below that Dec 4 is selected in the picker, but the text field shows Dec 3. The console.log shows that the time is set to 12AM CST, which means when it converts to a pacific date in this mixin function, it sets the date back a day.

Screenshot 2023-12-06 at 8 51 08 AM

My solution is to update this function to allow the option of ignoring the timezone.

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [reply src/components/date-picker/DatePicker.vue:20]: Did you see this? Does it help? https://stackoverflow.com/questions/75959657/replacement-for-vue-2-v-on-in-vue-3

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [inline src/components/date-picker/DatePicker.vue:33]: This is a two-way binding now, right? So does displayDate get updated when the user types in something?

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [inline src/components/date-picker/DatePicker.vue:170]: This is a significant change to what was emitted previously, which would require some rework of all UIs that use this shared component.

I recommend emitting the same string value as before.

Comment (severinbeauvais) [reply src/components/date-picker/DatePicker.vue:170]: OK, I'm a bit confused -- what's the difference between emitDate() above and emitDateSync() here -- why do we need both?

File: src/components/date-picker/package.json
Comment (severinbeauvais) [inline src/components/date-picker/package.json:9]: I think you don't actually import the shared mixins. Line 84 in the component imports the "local mixin" -- a relative path from this component. (`@/mixins` always resolves to the local mixins folder.)

Basically, when this component is used in Storybook, it uses DateMixin in this repo's mixins folder, but when this component is used in an app, it uses DateMixin in the app's mixins folder.

See if you can get this working while importing the shared mixin instead. 🤞

File: src/mixins/date-mixin.ts
Comment (severinbeauvais) [reply src/mixins/date-mixin.ts:72]: Two things:
1. I'd rather not change the date mixin.
2. When this component is used in an app, it's using the app's date mixin (see comment above).

If this method isn't working for you then it means the `date` parameter may be invalid. Normally, a date object is UTC/GMT internally and has a tz for display purposes. Try some of these on `date` and see how it's set internally.

image

As for this method, it should always convert a date into America/Vancouver time zone.

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [reply src/components/date-picker/DatePicker.vue:170]: Good point, I'll emit the string instead of the Date. I'm not too sure what the point of emitDateSync() is. I'll dig through some of the UIs a bit to see if it's being used.

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [reply src/components/date-picker/DatePicker.vue:33]: It doesn't update since the field is readonly. The picker wasn't updating when it was `value`, I think because in Vue 3 they changed that prop to `model-value`: https://v3-migration.vuejs.org/breaking-changes/v-model.html

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [reply src/components/date-picker/DatePicker.vue:170]: It seems it's used in [Edit UI](https://github.com/bcgov/business-edit-ui/blob/main/src/dialogs/ResolutionDateDialog.vue#L22) and a [few places in Create UI](https://github.com/search?q=repo%3Abcgov%2Fbusiness-create-ui+emitDateSync&type=code)

File: src/components/date-picker/package.json
Comment (leodube-aot) [reply src/components/date-picker/package.json:9]: Tried importing the shared mixin and the component seems to work fine. I'll include in my next commit.

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [reply src/components/date-picker/DatePicker.vue:33]: The way I've seen it done in a couple of Ketaki's tickets is to use kebab-case, I think because we generally use it for our props already

File: src/mixins/date-mixin.ts
Comment (severinbeauvais) [inline src/mixins/date-mixin.ts:68]: This is key information. Please restore it.

(The resulting yyyy-mm-dd may be different in a different timezone.)

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [inline src/components/date-picker/DatePicker.vue:150]: I think you should use the DateMixin (or DateUtilities) function to convert this.

The problem is that a UTC/GMT/ISO string is 7 or 8 hours ahead of Pacific time, so you will get the wrong value here for the first 7 or 8 hours of a day (it will be ahead by 1 day).

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [reply src/components/date-picker/DatePicker.vue]: So both are needed? Please resolve if no change is needed.

**Comment by severinbeauvais** [comment]
> > Does it work if you import the "local" mixin instead of the shared one?
> 
> No unfortunately it doesn't

OK, that's partly good news. (It means the previous issue was resolved.)

What if you add the shared mixin to Storybook's (ie, this repo's) package file?

**Comment by severinbeauvais** [comment]
> > What if you add the shared mixin to Storybook's (ie, this repo's) package file?
> 
> It looks like the shared mixin is already a dependency of the base package.json

Oh yeah. So maybe it's a Storybook build/config issue? How does Storybook build this so that DatePicker can see the mixins?

@ketaki-deodhar , have you run across this yet?

**Comment by ketaki-deodhar** [comment]
> > > What if you add the shared mixin to Storybook's (ie, this repo's) package file?
> > 
> > 
> > It looks like the shared mixin is already a dependency of the base package.json
> 
> Oh yeah. So maybe it's a Storybook build/config issue? How does Storybook build this so that DatePicker can see the mixins?
> 
> @ketaki-deodhar , have you run across this yet?

@severinbeauvais not really. Let me take a look

File: src/components/date-picker/DatePicker.vue
Comment (leodube-aot) [reply src/components/date-picker/DatePicker.vue]: You're right, it would've been the wrong value early in the day. I've replaced the `toISOString` calls I used with functions from DateMixin. However, using the mixin still results in the previous problem of the displayed date being wrong since the picker defaults to 00:00:00 GMT time which, when converted to PST by the mixin, puts the displayed value back a day.

I think the previous solution of using `toLocaleDateString` without converting the timezone would work, but I'm not sure if it should go in the util or in the component itself.

**Comment by leodube-aot** [comment]
> can you try passing the mixin in the mount? Not sure if it will work or if right way to do it but something like below..... !

No luck unfortunately. I passed the mixin to the mount like so:

Screenshot 2023-12-07 at 2 55 02 PM

I tried it with the previous method of adding the mixin to the component, which doesn't seem to register the mixin based on the error message. In this case, both storybook and the unit tests fail.

Screenshot 2023-12-07 at 2 52 44 PM

I also tried it with the way I have it in this PR. In this case storybook works but the unit tests fail.

Screenshot 2023-12-07 at 2 53 41 PM

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [reply src/components/date-picker/DatePicker.vue]: Oh, I didn't realize the picker sets it to 0Z :(

As this is specific to this component, I think you should work around it in here. My suggestion is to create a computed (local getter) in here to provide a yyyy-mm-dd that you expect (ie, yes, truncate the ISO date if that's easiest). Then you can use this computed to emit the events.

Comment (leodube-aot) [reply src/components/date-picker/DatePicker.vue]: I've fixed it following this approach of using a local getter and I think it it works pretty well, even if it's partially duplicated from the mixin.

**Comment by leodube-aot** [comment]
> OK, you got it down to a unit test error. That's progress. And if we decide to give up, we can forgo the unit tests for now (but we'd create a Tech Debt ticket to fix them later).

It also fails before any of the tests are run (see the test summary copied below, I'm running the tests only for DatePicker.spec.ts so that's why there's only one test file). I imagine it might be some issue with the test setup that's thrown from the component. One thing I'm confused by is the `mount`ing of the component only occurs within the tests themselves, but based on the output below none of them get a chance to run. 

```
Test Files  1 failed (1)
Tests  no tests
Start at  15:14:26
Duration  2.06s
```

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [inline src/components/date-picker/DatePicker.vue:131]: Please add a comment indicating you need to do this because the picker returns a Date object with 00:00 UTC.

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [reply src/components/date-picker/DatePicker.vue]: Looks good to me. Pretty clean!

File: src/components/date-picker/DatePicker.vue
Comment (severinbeauvais) [reply src/components/date-picker/DatePicker.vue:136]: Perfect 👍

**Review by severinbeauvais — APPROVED** [review]

**Comment by leodube-aot** [comment]
Created new techdebt ticket for fixing up the unit tests. I'll get one more review before merging this PR.

https://app.zenhub.com/workspaces/entities-team-space-6143567664fb320019b81f39/issues/gh/bcgov/entity/18943

**Review by ketaki-deodhar — APPROVED** [review]
Looks good!

**Comment by ketaki-deodhar** [comment]
> > can you try passing the mixin in the mount? Not sure if it will work or if right way to do it but something like below..... !
> 
> No luck unfortunately. I passed the mixin to the mount in the unit tests like so:
> 
> Screenshot 2023-12-07 at 2 55 02 PM
> I ran it with the previous method of adding the mixin to the component, which doesn't seem to register the mixin based on the error message. In this case, both storybook and the unit tests fail.
> 
> Screenshot 2023-12-07 at 2 52 44 PM
> I also tried running it with the way I extend the mixin in this PR. In this case storybook works but the unit tests fail.
> 
> Screenshot 2023-12-07 at 2 53 41 PM

Unit tests are a big pain point :(

**Comment by leodube-aot** [comment]
> Unit tests are a big pain point :(

I agree haha I felt like I was going in circles yesterday

**Pull request merged**
